### PR TITLE
New documentation draft:

### PR DESCRIPTION
**Pull request by xsawyerx**
Docs are hard. We currently have the Introduction, the Tutorial,
the Manual, and the Cookbook. It's hard to tell what is documented
where.

The idea was that the Introduction explains things gently,
the Tutorial walks you through writing an application, the Manual
provides the complete coverage, and the Cookbook provides all the
tips, tricks, and hints collected by developers (based heavily
on contributor input).

I think the problem is mainly that we do not have a base-line for
what skills you need in order to understand Dancer and how to use
it. This means we have multiple places to start from, but they
are not clear. Each one tries to cater to multiple different
crowds.

What @mickeyn and I wanted to create was a simple guide that,
assuming you know basic HTTP and Perl, explains the structure of
the Dancer system: the Apps, the routing, the hooks, etc.

Starting from the ground up, this documentation is relatively
thin but is able to not only cover a large and significant portion
of the keywords, it also captures how Dancer works, and how its
structured internally (to an extent).

### CODE REVIEW

**Comment by xsawyerx** [comment]
@PerlDancer, any comments?

**Comment by racke** [comment]
Not yet, will check in later!

**Comment by SysPete** [comment]
excellent :+1:

**Comment by SysPete** [comment]
@racke you promised to have a look at this...

**Comment by racke** [comment]
Thanks for the reminder, @SysPete. Let's start with the simple things first - I found a small typo: https://github.com/racke/dancer2/commit/8c58764f62f601441c37263b09c3368968df58af.

Is there a way to get HTML out of the PODs with my old friend Dist::Zilla?

Simply using Pod::Simple::HTMLBatch won't work with the special POD elements.

**Comment by xsawyerx** [comment]
@racke++ Thanks for the typo fix. I missed it.

Basically, `dzil build` and `perldoc` the POD. Wish it was easier, but it's not that bad, I guess.

**Comment by racke** [comment]
The problem with perldoc is that it doesn't show any links etc.

**Comment by racke** [comment]
get
Handles GET methods.

I think this should say "Handles GET method" - we have only one.

**Comment by racke** [comment]
It had a first glance at it before and will continue to review it. It is definitely well written and a big improvement for any user, especially new users.

**Comment by racke** [comment]
What about adding some suggestions about common plugins? So many users try to reinvent the wheel for authentication, database access, you name it.

**Comment by xsawyerx** [comment]
**GET methods** and **GET method** are both wrong, IMHO. What I should have written was **handles requests using GET method**.

Common plugins is a good idea, except the list of recommended modules would be limited. We can also half-endorse. Like, **this exists**.

Another note to self: would be great to have a synopsis including all the core pieces.

**Comment by racke** [comment]
This documentation doesn't go into the details about the engines - which is fine, but we should add pointers
from the Engine section so users can go into the details for loggers etc.

**Comment by xsawyerx** [comment]
What details would you like it to go into?

**Comment by racke** [comment]
Like an overview on the existing loggers and the configuration options.

**Comment by racke** [comment]
It would probably be easier to merge this draft and add new issues for the stuff mentioned here :-).

**Comment by xsawyerx** [comment]
Should we ask on the mailing list for people to review this?

What's hard for me is to answer the question: Is this better than what we had before? To me it feels like it because it's shorter, succinct, and (IMHO) teaches you more than the previous documentation - but I'm not sure what other people think.

**Comment by racke** [comment]
On 03/08/2016 09:26 AM, Sawyer X wrote:
> Should we ask on the mailing list for people to review this?
> 
> What's hard for me is to answer the question: Is this better than what we had before? 
> To me it feels like it because it's shorter, succinct, and (IMHO) teaches you more than the previous documentation - but I'm not sure what other people think.
> 

I'm quite convinced that this is better than before - and it can't hurt to give the people a chance to review it.
I suggest to put in somewhere as HTML so people can browse it easily.

Regards
	Racke


-- 
Perl and Dancer Development

File: lib/Dancer2.pm [hunk 1]
```diff
@@ -115,102 +115,1022 @@ __END__
...
+
 =head1 DESCRIPTION
 
-Dancer2 is the new generation of L<Dancer>, the lightweight web-framework for
-Perl. Dancer2 is a complete rewrite based on L<Moo>.
+Dancer2 is a lightweight micro web framework for Perl. It is the new
+generation of L<Dancer> and replaces L<Dancer>.
+
+If you are converting a L<Dancer> application to L<Dancer2>, the
+L<Dancer2::Manual::Migration> document covers the changes you might
```
Comment (jasonblewis) [inline lib/Dancer2.pm]: "might" sounds a bit vague - consider deleting "might"

Comment (xsawyerx) [inline lib/Dancer2.pm]: I'm not sure every app actually needs updating. I don't want to make an assumption about it.

Comment (racke) [inline lib/Dancer2.pm]: On 04/07/2016 01:36 PM, Sawyer X wrote:
> In lib/Dancer2.pm <https://github.com/PerlDancer/Dancer2/pull/1069#discussion_r58856895>:
> 
>>  =head1 DESCRIPTION
>>  
>> -Dancer2 is the new generation of L<Dancer>, the lightweight web-framework for
>> -Perl. Dancer2 is a complete rewrite based on L<Moo>.
>> +Dancer2 is a lightweight micro web framework for Perl. It is the new
>> +generation of L<Dancer> and replaces L<Dancer>.
>> +
>> +If you are converting a L<Dancer> application to L<Dancer2>, the
>> +L<Dancer2::Manual::Migration> document covers the changes you might
> 
> I'm not sure every app actually needs updating. I don't want to make an assumption about it.

At the very least you need to replace Dancer with Dancer2 :-).

Regards
          Racke


-- 
Perl and Dancer Development

**Comment by jasonblewis** [comment]
Maybe something along the lines of:

To convert a `L<Dancer>` application to `L<Dancer2>`, it may simply be a matter of changing `use Dancer;` to `use Dancer2;`. The `+L<Dancer2::Manual::Migration>` document covers any other changes you might...

File: lib/Dancer2.pm (hunk 1 above)
Comment (xsawyerx) [inline lib/Dancer2.pm]: You're right. I'll make that correction.

**Comment by xsawyerx** [comment]
I like that suggestion very much! I'll use that.

**Comment by xsawyerx** [comment]
Updated.

**Comment by xsawyerx** [comment]
What isn't discussed here yet is deprecating the existing Manual. This should effectively replace it. We would still keep the deployment doc, the migration doc, and an index of all the DSL. For now we can also keep the tutorial, although that can (and probably should) be split to another distribution.

What says @PerlDancer team?

**Comment by racke** [comment]
I find the example code in the tutorial not very enticing, also it may lead people on the wrong path (code on their own instead using DBIC / DPAE). So I'm fine with removing that as well.

Otherwise I agree with the deprecation.

**Comment by cromedome** [comment]
I agree with @racke.

**Comment by xsawyerx** [comment]
Okay I want to push this forward. I realized we still have the old documentation in and I keep pointing to these commits as a more readable documentation. What is left on this to just merge it?
(I rebased the branch.)

**Comment by cromedome** [comment]
@xsawyerx I was really fond of this rewrite. I am a bit embarrassed to say that I haven't looked at them in a while though, and will take another read through once I finish this release this week. Then, let's please come up with a plan for merging and pushing these out, and deprecating the old Manual 💃  Will that work?

**Comment by xsawyerx** [comment]
Here's an updated list of what this branch still needs to have in order to remove `Dancer2::Manual`:

- [ ] Document the rest of the hooks
- [ ] Document delayed responses
- [ ] File Handlers
- [ ] Moving the documentation on writing your own handler somewhere else?
- [ ] Error handling
- [ ] `send_file`
- [ ] Static file handling
- [ ] Uploading
- [ ] Configuration
- [ ] Logging
- [ ] Testing
- [ ] Packaging
- [ ] Middlewares
- [ ] Plugins
- [ ] DSL keywords

So, basically, this is tough. We can't merge this yet. I'm not sure one documentation is good for all of this. Perhaps it's time to break this up.

**Comment by bigpresh** [comment]
I'd be inclined to say that, if the merge conflicts can be fixed, this ought to be merged, and the doco as it then stands re-reviewed and improved further, using your checklist for inspiration.  Even if it's not 100% complete, it looks to be a good improvement.

If I get a chance soon, I'd be happy to sit down and review the doco in detail and see what improvements I can come up with, too.

**Comment by bigpresh** [comment]
Right - this one needs addressing - good doco is incredibly important, and the great work done in this PR is sitting unused!  I'm of the opinion that, even if there's some improvements still to be made, it's a good improvement, and so we should merge the bloody thing then improve iteratively, otherwise it'll never get in. Thoughts?

**Comment by xsawyerx** [comment]
I only have two problems: 1. I don't want to merge this and to lose documentation for stuff I listed. 2. I want additional documentation to take the ideas I had in mind when writing this.

I can review this again over the weekend and try to add what's missing - at least the big stuff. What would be great is if it I can get around to write an internal documentation on how to document, like a style guide.

**Comment by xsawyerx** [comment]
I can rebase this branch as well.

**Comment by ambs** [comment]
Please take a decision on this branch. Having it here so much time doesn't do any good. More like bad. Rebase it, and make your list of enhancements a list of issues, so people can pick low hanging fruit.

**Comment by sdondley** [comment]
I'd like to pitch and help where I can. I'm totally new to Dancer. However, this can be an asset as I can help point out what might be confusing to a newb. I'm a pretty decent writer.

And I've never used GitHub and git on a serious project so it's a bit disorienting for me right now. So a little patience would be appreciated. Once I get my bearings, I should be able to help quite a bit.

**Comment by cromedome** [comment]
I agree with @ambs - anything we can do to get this going again and merged would be great. People have a love/hate relationship with the docs it seems, so doing something to make the docs even better would be a welcome improvement!

**Comment by sdondley** [comment]
Not having any direction, I went ahead and committed some changes last night to Manual.pod, mostly as an exercise for myself to help solidify what I already know about Dancer2 and just to help get a better overall feel for the current quality of the documentation as it currently stands. See #1424 for my pull request. Feel free to incorporate, or not, any of my changes as you see fit.

Before I go too much further, though, I'd like to get a solid commitment that others are still interested in working collaboratively on this task which is now over 2 years old and doesn't show much signs of life. Can we set some kind of arbitrary deadline, like say, the end of January to have this effort wrapped up? And who, exactly, is still actively interested in working on this? Without knowing, I feel a bit left up in the air.

**Comment by xsawyerx** [comment]
@cromedome and I discussed this work and we formed a plan for moving forward on it:

- [X] Go over the docs and freshen them up a bit
- [ ] Go over the docs again and include all the comments from the discussion in this ticket (I have those listed)
- [ ] Go over the docs again and include all the enhancements we have done since this PR was created (I have those listed)
- [ ] Go over the topics I listed in an earlier comment of mine on this ticket and add them to the documentation
- [ ] Go over the existing `Dancer2::Manual` double-check everything is covered
- [ ] Replace `Dancer2::Manual` with this and create an index in `Dancer2.pm` (see section below)
- [ ] Request feedback from core developers and documentation contributors
- [ ] Request feedback on the mailing list
- [ ] Merge and release!

Additionally, we are thinking of cleaning up the documentation paths:

- [ ] `Dancer2.pm` will become the index, pointing to all the manual sections
- [ ] `Dancer2::Manual` will host the new documentation that this PR shows (right now it's in `Dancer2.pm` but I'll eventually move it to `Dancer2::Manual`
- [ ] `Dancer2::Tutorial` will likely need to be rewritten and renamed to `Dancer2::Manual::Tutorial`
- [ ] `Dancer2::Cookbook` will be renamed to `Dancer2::Manual::Cookbook`
- [ ] `Dancer2::Manual::Keywords` will be renamed to `Dancer2::Manual::Reference` (not 100% on this one yet)

File: lib/Dancer2.pm [hunk 2]
```diff
@@ -1002,6 +1029,50 @@ content as the response:
         template 'index'; # render views/index.tt
     };
 
+In effect, the result type is determined by two factors: The content type
+the application returns (C<text/html> by default) and the returned string
+(which, if you're using C<template>, will be an HTML template in your
+F<views/> directory).
```
Comment (racke) [inline lib/Dancer2.pm:1035]: I think that is really confusing.

**Comment by xsawyerx** [comment]
@racke not sure the fixes are good. Feel free to adjust. I couldn't come up with a better way to deal with some of them.

File: lib/Dancer2.pm [hunk 3]
```diff
@@ -121,82 +121,1018 @@ __END__
...
+    get qr{^/view/$} => sub {...};
+
+You will need to anchor your regular expressions yourself, in order to
+allow you to provide path segments in regular expressions too.
+
+    get qr{/view/} => sub {...};
+
+This will match both F</view/> and F</new/view/here>.
+
+Any capturing done in regular expressionos are available using the
```
Comment (cromedome) [inline lib/Dancer2.pm]: `s/expressionos/expressions/;`

Comment (xsawyerx) [reply lib/Dancer2.pm]: I can't find this in the most recent version.
